src/types.py

```diff
@@ -117,64 +132,153 @@ def is_int(self) -> bool:
+            storage=[],
+            funcspec=[],
+            init=None,
+            bitsize=None,
+        )
+        set_decl_name(decl)
+        return to_c(decl)
 
-    def to_decl(self, var: str, fmt: Formatter) -> str:
-        ret = self.format(fmt)
-        prefix = ret if ret.endswith("*") else ret + " "
-        return prefix + var
+    def _to_ctype(self, seen: Set["Type"], fmt: Formatter) -> CType:
+        def simple_ctype(typename: str) -> ca.TypeDecl:
+            return ca.TypeDecl(
+                type=ca.IdentifierType(names=[typename]), declname=None, quals=[]
+            )
 
-    def _stringify(self, seen: Set["Type"], fmt: Formatter) -> str:
         unk_symbol = "MIPS2C_UNK" if fmt.valid_syntax else "?"
+
         if self in seen:
-            return unk_symbol
+            return simple_ctype(unk_symbol)
         seen.add(self)
         type = self.get_representative()
         size = type.size or 32
         sign = "s" if type.sign & Type.SIGNED else "u"
-        if type.kind in (Type.K_ANY, Type.K_ANYREG):
+
+        if (type.kind & Type.K_ANYREG) == Type.K_ANYREG:
             if type.size is not None:
-                return f"{unk_symbol}{size}"
-            return unk_symbol
+                return simple_ctype(f"{unk_symbol}{size}")
+            return simple_ctype(unk_symbol)
+
+        if type.kind == Type.K_FLOAT:
+            return simple_ctype(f"f{size}")
+
         if type.kind == Type.K_PTR:
             if type.ptr_to is None:
-                return "void *"
-            ctype = type.get_pointer_to_ctype()
-            if ctype is not None:
-                return type_to_string(ca.PtrDecl(quals=[], type=ctype))
-            return (type.ptr_to._stringify(seen, fmt) + " *").replace("* *", "**")
-        if type.kind == Type.K_FLOAT:
-            return f"f{size}"
+                return ca.PtrDecl(type=simple_ctype("void"), quals=[])
+            return ca.PtrDecl(type=type.ptr_to._to_ctype(seen, fmt), quals=[])
+
         if type.kind == Type.K_CTYPE:
             if type.ctype_ref is None:
-                return "?"
-            return type_to_string(type.ctype_ref)
-        return f"{sign}{size}"
+                return simple_ctype(unk_symbol)
+            return type.ctype_ref
+
+        if type.kind == Type.K_FN:
+            if type.fn_sig is None:
+                return simple_ctype(unk_symbol)
+            return_ctype = type.fn_sig.return_type._to_ctype(seen.copy(), fmt)
+
+            params: List[Union[ca.Decl, ca.ID, ca.Typename, ca.EllipsisParam]] = []
+            for param in type.fn_sig.params:
+                decl = ca.Decl(
+                    name=param.name,
+                    type=param.type._to_ctype(seen.copy(), fmt),
+                    quals=[],
+                    storage=[],
+                    funcspec=[],
+                    init=None,
+                    bitsize=None,
+                )
+                set_decl_name(decl)
+                params.append(decl)
```

Comment:
I'm going to stick this on the `K_CTYPE `branch, so that `_to_ctype(...)` always returns a clean copy of `CType` that can be modified.

The function parameters do have names by default, the names are manually stripped out by `type_from_global_decl` when parsing the `TypeMap.var_types` dict. The types are *not* stripped out from the `TypeMap.functions` dict.